lib/charms/vault_k8s/v0/vault_kv.py

```diff
@@ -0,0 +1,682 @@
+        for request in kv_requests:
+            if not self._credentials_issued_for_request(
+                nonce=request.nonce, relation_id=request.relation_id
+            ):
+                outstanding_requests.append(request)
+        return outstanding_requests
+
+    def get_kv_requests(self, relation_id: int | None = None) -> List[KVRequest]:
+        """Get all KV requests for the relation."""
+        kv_requests: List[KVRequest] = []
+        relations = (
+            [
+                relation
+                for relation in self.model.relations[self.relation_name]
+                if relation.id == relation_id
+            ]
+            if relation_id is not None
+            else self.model.relations.get(self.relation_name, [])
+        )
+        for relation in relations:
+            assert isinstance(relation.app, ops.Application)
+            if not relation.active:
+                continue
+            app_data = relation.data[relation.app]
+            for unit in relation.units:
+                unit_data = relation.data[unit]
+                if not is_requirer_data_valid(app_data=app_data, unit_data=unit_data):
+                    continue
+                kv_requests.append(
+                    KVRequest(
+                        relation_id=relation.id,
+                        app_name=relation.app.name,
+                        unit_name=unit.name,
+                        mount_suffix=app_data["mount_suffix"],
+                        egress_subnets=get_egress_subnets_list_from_relation_data(unit_data),
+                        nonce=unit_data["nonce"],
+                    )
+                )
+        return kv_requests
+
+    def _credentials_issued_for_request(self, nonce: str, relation_id: int | None) -> bool:
+        """Return whether credentials have been issued for the request."""
+        relation = self.model.get_relation(self.relation_name, relation_id)
+        if not relation:
+            return False
+        credentials = self.get_credentials(relation)
+        return credentials.get(nonce) is not None
+
+
+class VaultKvConnectedEvent(ops.EventBase):
+    """VaultKvConnectedEvent Event."""
+
+    def __init__(
+        self,
+        handle: ops.Handle,
+        relation_id: int,
+        relation_name: str,
+    ):
+        super().__init__(handle)
+        self.relation_id = relation_id
+        self.relation_name = relation_name
+
+    def snapshot(self) -> dict:
+        """Return snapshot data that should be persisted."""
+        return {
+            "relation_id": self.relation_id,
+            "relation_name": self.relation_name,
+        }
+
+    def restore(self, snapshot: Dict[str, Any]):
+        """Restore the value state from a given snapshot."""
+        super().restore(snapshot)
+        self.relation_id = snapshot["relation_id"]
+        self.relation_name = snapshot["relation_name"]
+
+
+class VaultKvReadyEvent(ops.EventBase):
```

Comment:
Can you extract this logic and the one from `VaultKvConnectedEvent` to a superclass to avoid repeating the code?

Reply:
done in #499